vm/systemSmartContracts/staking.go

```diff
@@ -0,0 +1,226 @@
+	if args.CallValue.Cmp(r.stakeValue) != 0 {
+		return vmcommon.UserError
+	}
+
+	registrationData := stakingData{
+		StartNonce:    0,
+		Staked:        false,
+		BlsPubKey:     nil,
+		UnStakedNonce: 0,
+		StakeValue:    big.NewInt(0),
+	}
+	data := r.eei.GetStorage(args.CallerAddr)
+
+	if data != nil {
+		err := json.Unmarshal(data, registrationData)
+		if err != nil {
+			log.Error("unmarshal error on staking smart contract stake function " + err.Error())
+			return vmcommon.UserError
+		}
+	}
+
+	if registrationData.Staked == true {
+		log.Error("account already staked, re-staking is invalid")
+		return vmcommon.UserError
+	}
+
+	registrationData.Staked = true
+
+	if len(args.Arguments) < 1 {
+		log.Error("not enough arguments to process stake function")
+		return vmcommon.UserError
+	}
+
+	registrationData.StartNonce = args.Header.Number.Uint64()
+	registrationData.BlsPubKey = args.Arguments[0].Bytes()
+	//TODO: verify if blsPubKey is valid
+
+	data, err := json.Marshal(registrationData)
+	if err != nil {
+		log.Error("marshal error on staking smart contract stake function " + err.Error())
+		return vmcommon.UserError
+	}
+
+	r.eei.SetStorage(args.CallerAddr, data)
+
+	err = r.eei.Transfer(args.RecipientAddr, args.CallerAddr, args.CallValue, nil)
+	if err != nil {
+		log.Error("transfer error on stake function " + err.Error())
+	}
+
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) unStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	var registrationData stakingData
+	data := r.eei.GetStorage(args.CallerAddr)
+	if data == nil {
+		log.Error("unStake is not possible for address which is not staked")
+		return vmcommon.UserError
+	}
+
+	err := json.Unmarshal(data, registrationData)
+	if err != nil {
+		log.Error("unmarshal error in unStake function of staking smart contract " + err.Error())
+		return vmcommon.UserError
+	}
+
+	registrationData.Staked = false
+	registrationData.UnStakedNonce = args.Header.Number.Uint64()
+
+	data, err = json.Marshal(registrationData)
+	if err != nil {
+		log.Error("marshal error in unStake function of staking smart contract" + err.Error())
+		return vmcommon.UserError
+	}
+
+	r.eei.SetStorage(args.CallerAddr, data)
+
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) finalizeUnStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	ownerAddress := r.eei.GetStorage([]byte(ownerKey))
+	if !bytes.Equal(ownerAddress, args.CallerAddr) {
+		return vmcommon.UserError
+	}
+
+	var registrationData stakingData
+	for _, arg := range args.Arguments {
+		data := r.eei.GetStorage(arg.Bytes())
+		err := json.Unmarshal(data, registrationData)
+		if err != nil {
+			log.Error("unmarshal error on finalize unstake function" + err.Error())
+			return vmcommon.UserError
+		}
+
+		if registrationData.UnStakedNonce == 0 {
+			log.Error("validator did not unstaked yet")
+			return vmcommon.UserError
+		}
+
+		r.eei.SetStorage(arg.Bytes(), nil)
```

Comment:
Might also add a check here to not be able to call finalize un stake before n blocks passed.

Reply:
These things are saved in a local cache, so if you return with userError, non of the cached values are taken into consideration, TxValue is sent back to the sender, gas is consumed. That code is in SmartContract processor.